### PR TITLE
ES6 Bootstrap

### PR DESCRIPTION
**Pull request by hsab**
## ES6 Bootstrap
This pull request follows the #3758 discussion; with the aim of transitioning p5 to ES6. It consists of minor changes to the build system to facilitate processing, linting and testing the library with ES6 syntax and standards, as well as major and ubiquitous syntactical modification in line with ES6 features.

It is worthy to note that these transformations are by no means complete, and do not reflect nor implement every possible feature of ES6. They are intended to facilitate a smoother transition to properly and efficiently utilize ES6 features if and when aligned with the community interests and standards. And serve to motivate contributors to gradually conform to the new style and features.

These changes are applied with the following in mind:
- Reduce syntactical ambiguity and JS idiosyncrasies.
- Increase clarity and refrain from obfuscation resulting from new features (e.g.  [Destructuring](http://es6-features.org/#ParameterContextMatching))
 - Prioritize performance when necessary (e.g. [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach))

Moreover, there are many areas of p5 that could benefit from ES6 features beyond the scope of this PR. These features include but are not limited to: 
- Generators
- Number Type Checking
- Set/Map Data-Structures
- Typed Arrays
- etc.

Unfortunately, implementing, testing and benchmarking all these features, was beyond my availability, expertise, and code familiarity.

### Process
After modifying the build system to accommodate the migration, a series of transformations are applied, bootstrapped by [lebab](https://github.com/lebab/lebab), an ES5 to ES6/ES7 transpiler. For every transformation the following process was performed:
- Apply transformation using lebab
- Ensure correct formatting using `npm run lint:fix`
- Manually inspect every single line that was modified and fix/adjust if needed
- Test and build using `npm run grunt`
- Check the example(s) of affected function(s) in the reference doc
- Commit the transformation

In the end, the library was tested against a large random selection of the examples available through [p5.js-web-editor](https://github.com/processing/p5.js-web-editor/) to ensure correct functionality. This was done by cloning the web editor repository and using the generated library files (`p5.js` and `p5-min.js`) instead of the CDN provided 0.8.0 version.

### Issues & Concerns:

- `new p5()` fails to execute when using `combineModules`, although global mode functions as expected. The error thrown is `p5 is not a constructor`
- Older browsers have not been tested given that the library is built against [browserlist](https://github.com/browserslist/browserslist) `last 2 versions` and `not dead`. ( https://github.com/processing/p5.js/issues/3758#issuecomment-496296302 )
- Functionality has been only tested on Chrome and Firefox

### Build System and General Changes
- Every instance of `require` has been changed to import (both node, e.g. `fs`, and `p5` modules)
- Constants are imported such that syntax replicates old format: `constants.TWO_PI`
- Every module is exported in ES6 style
  Including `src/core/helpers.js` and `src/core/constants.js`
  With the exception of `app.js` still using `module.exports = p5;`
- Using `@babel/register` to properly perform `mochaTest:test` suite with babel as a compiler
- Replaced [`brfs`](https://github.com/browserify/brfs) with [`babel-plugin-static-fs`](https://www.npmjs.com/package/babel-plugin-static-fs)
   `brfs` is completely removed as a dependency
   `static-fs` allows for `import { readFileSync } from 'fs';`
- Fixed `combineModules.js` to properly handle the migration to `import` syntax
  **Issue:** `p5 is not a constructor` when using the custom bundle in instanced mode

### Summary of Transformations
The first half of these commits, from https://github.com/processing/p5.js/commit/1375a4385445f1b21ccfa665e758fdff1deca586 to https://github.com/processing/p5.js/commit/0733e01f692ba515ab1877f04bf2d61de7b6c6db tackles the build system to properly process the transition. 

The second half of these commits, from https://github.com/processing/p5.js/commit/7ad6a4da68b0f0511ce4085dbc9359187a9ffe45 to https://github.com/processing/p5.js/commit/054f29127142b15377bcfd09952ab93a1609f1e8 conforms the library to a limited set of ES6 features:

- [Modules: Export/Import](http://es6-features.org/#ValueExportImport): https://github.com/processing/p5.js/commit/07f315db03c6aa7a60221ae8a399af48150183df https://github.com/processing/p5.js/commit/bc612039d2e82ef0c991efd0a54e13bd06334697 https://github.com/processing/p5.js/commit/7d1d26bb3fa1376867e44e3c629325aa5ae502ce https://github.com/processing/p5.js/commit/35c54330e307a2fd7653346e1bb2f9faf72d3477 https://github.com/processing/p5.js/commit/0791e3b7f2c57386d4c46048b35a8520cf8fdaf9 https://github.com/processing/p5.js/commit/361907e822dc0764a0f085a3bb95667c1fc2ca8f 
- [Modules: Default & Wildcard](http://es6-features.org/#ValueExportImport): https://github.com/processing/p5.js/commit/d898fc5122108bec156f0d93eac2f8aa118ae4b0

- [Constants](http://es6-features.org/#Constants): https://github.com/processing/p5.js/commit/7ad6a4da68b0f0511ce4085dbc9359187a9ffe45 https://github.com/processing/p5.js/commit/f57bd77a370ef87011bd675e913bc6204c7692c1

- [Block-Scoped Variables](http://es6-features.org/#BlockScopedVariables): https://github.com/processing/p5.js/commit/7ad6a4da68b0f0511ce4085dbc9359187a9ffe45 https://github.com/processing/p5.js/commit/f57bd77a370ef87011bd675e913bc6204c7692c1

- [Rest Parameter](http://es6-features.org/#RestParameter): https://github.com/processing/p5.js/commit/c6329d7c8c7e5693a5ab4842922fb7f8d574da88 

- [Spread Operator](http://es6-features.org/#SpreadOperator): https://github.com/processing/p5.js/commit/b4e4b1ffd6b6878bd248a5ea7ac7d9d86ea65e78

- [Method Properties](http://es6-features.org/#MethodProperties): https://github.com/processing/p5.js/commit/70d7e5484940ea2383343ac9b99baa02b8114d22

- [Property Shorthand](http://es6-features.org/#PropertyShorthand): https://github.com/processing/p5.js/commit/9693c847e21f691db26a0d3fa72e955f0ad6b59f

- [Arrow Functions: Statement Bodies](http://es6-features.org/#StatementBodies): https://github.com/processing/p5.js/commit/694177b46055032c864feed8f2941e2ab0726824

- [Arrow Functions: Expression Bodies](http://es6-features.org/#ExpressionBodies): https://github.com/processing/p5.js/commit/95274cb0205ebd866bbd5aeffd8ba33619265bbd

- [Class Definition](http://es6-features.org/#ClassDefinition): https://github.com/processing/p5.js/commit/c97cc486fb67687b614cb2ff795ac05926b7956b

- [Class Inheritance](http://es6-features.org/#ClassInheritance): https://github.com/processing/p5.js/commit/c97cc486fb67687b614cb2ff795ac05926b7956b

- [Template Literals: String Interpolation](http://es6-features.org/#StringInterpolation): https://github.com/processing/p5.js/commit/b51a8a4984e03e980772b9854d2a46ced5ce63e3

- [Template Literals: Custom Interpolation](http://es6-features.org/#CustomInterpolation): https://github.com/processing/p5.js/commit/b51a8a4984e03e980772b9854d2a46ced5ce63e3

- [Default Parameter Values](http://es6-features.org/#DefaultParameterValues): https://github.com/processing/p5.js/commit/e66ff6baba776c60cb0945fead2c4914b430b7eb

- [String Searching: Includes](http://es6-features.org/#StringSearching): https://github.com/processing/p5.js/commit/49bf914192f863ea8717f676ee74765c535633b4
- [Array Searching: Includes (ES7)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes): https://github.com/processing/p5.js/commit/49bf914192f863ea8717f676ee74765c535633b4

- [Iterators: For-Of](http://es6-features.org/#IteratorForOfOperator): https://github.com/processing/p5.js/commit/054f29127142b15377bcfd09952ab93a1609f1e8

### CODE REVIEW

**Comment by lmccart** [comment]
A few of us have reviewed this and I'm going to go ahead and merge. Since it's such a large change, I didn't want this to get messy and confusing with different threads of questions and comments. Others are welcome to leave comments on here or open new issues if you have them, and we can continue to break out issues as needed. 

We've already discussed the addition of a markdown file in the `developer_docs` folder that contains some of the content from @hsab's top post, so we have a record of the ES6 decisions made and future todos.

This is fantastic work @hsab. Thank you for leading this significant upgrade of p5.js!

**Comment by hsab** [comment]
Thank you all!
I'll probably won't be able to respond today, but will be available tomorrow and will also submit a PR for the dev docs. Hoping that this goes as smoothly as I expected it! :)

**Comment by limzykenneth** [comment]
Great work @hsab!

I have a particular question on the use of arrow functions to define prototype members. From what I can find it seems to affect the value of `this` in the member function (arrow function's `this` is defined when the function is evaluated, not when the object is created as such it doesn't point to the object instance created). For a discussion about it [here](https://stackoverflow.com/questions/31755186/es6-arrow-functions-not-working-on-the-prototype) and [here](https://stackoverflow.com/questions/34361379/are-arrow-functions-and-functions-equivalent-exchangeable). 

I can't identify with a quick look whether it is a problem here or not but if the use of arrow function on prototype member isn't really doing anything other than saving bytes it seems a bit superflous to me.

**Comment by hsab** [comment]
@limzykenneth Thanks!

This is certainly a valid and concerning point. I will have a thorough look tomorrow. 
Although I was aware of this behavior, I can't recall running into any issues with this. 
Also, just did a quick look and it seems like exactly what you mention is happening in [`main.js`](https://github.com/processing/p5.js/commit/694177b46055032c864feed8f2941e2ab0726824#diff-848b28a6f51426fffc2b0c6b183fa1c0). But these functions (`_start`, `_draw`, `_setup`) are critical (right?) and yet `this` seems to point to object instance during execution. This is quite a curious case. Definitely needs more investigation. If you have an idea of why, I would very much appreciate it.

**Comment by Spongman** [comment]
maybe it would be prudent to first carefully considering a list of changes that _should_ be made, instead of just applying a bunch of automatic refactorings and seeing what sticks?

**Comment by outofambit** [comment]
@Spongman if you have thoughts about the es6 refactoring, please open a new issue to discuss there. thanks!

**Comment by limzykenneth** [comment]
@hsab From my understanding, basically in the `main.js` case you pointed out, at the time the arrow function is evaluated, the value of `this` is the same as the value of `this._start`'s `this` since they are defined in the same scope and so share the same value for `this`. However, when attaching to the prototype, the value of `this` should be defined at the point of object creation (when `new Something()` is called) but with arrow functions, the value of `this` in the function will be defined where the function is defined.

So in simpler terms, the issue will be centered around the usage of arrow functions on prototype members, the rest should not encounter problems and in my opinion ok to use arrow functions.

**Comment by lmccart** [comment]
@Spongman these changes were proposed by @hsab in #3758, which included a link to an in-progress branch for review. There was ample time for everyone to weigh in. @hsab also discussed the individual decisions for specific ES6 functionality to incorporate with me, and several other mentors in this project. As mentioned in the process above, this was not an automatic refactor, every single line was checked by hand with the design decisions we'd discussed in mind. Each ES6 change has been cleanly broken out into its own commit, so if there is a particular aspect we want to revert or modify, this should not be too hard. We are open to feedback, anyone is welcome to open issues for specific points they'd like to discuss further, and we can go from there. Due to the nature of this major change, and the concern about ending up with incomplete pull requests or significant merge conflicts, we felt it would be easiest to get the bulk of the work merged, and fine-tune individual aspects afterward.

More generally, this sort of passive aggressive comment is unproductive and out of line with our [code of conduct](https://github.com/processing/p5.js/blob/master/CODE_OF_CONDUCT.md) and [community statement](https://p5js.org/community/). Our community includes many people that are new to coding, new to contributing to open source software, and/or people of underrepresented backgrounds that may often encounter exclusion, harassment, and other barriers to participation in OSS. We have made a commitment with p5 to maintain a community that is open and welcoming, prioritizing beginners and newbies, and supportive of learning.

I've unfortunately had a number of p5.js contributors mention to me that your posts over various threads have discouraged them and made them hesitant to participate. I have talked to you before about your comments and tone. I think it would be helpful if we could talk one on one over email to make sure we're both understanding each other, and the goals of this project. At this point, I must kindly ask that you refrain from posting on p5.js repositories until you and I have had that conversation. You can email me at laurenleemccarthy@gmail.com. Thank you.

**Comment by Spongman** [comment]
I’m sorry I don’t mean to sound aggressive. I was just concerned that some changes had been merged while there was still an open discussion about the performance impact of a couple of them. You had asked me to demonstrate those, and there was a subsequent discussion over the implications, but I had not seen a conclusion to that discussion, so I figured it was still open.

**Comment by sakshamsaxena** [comment]
Hey @hsab, for the below issue mentioned by you in description, could you maybe open a separate issue and mention me there ? I authored this wee little beast quite a long while ago, so maybe I can take a fresh look into it. Thanks! 
> new p5() fails to execute when using combineModules, although global mode functions as expected. The error thrown is p5 is not a constructor

**Comment by outofambit** [comment]
@sakshamsaxena it looks like you're already there, but for others: the issue for that is #3883 🙌
